src/App.tsx

```diff
@@ -1,24 +1,246 @@
+const ADDED_URL = `/todos?userId=${USER_ID}`;
 
 export const App: React.FC = () => {
+  const [todosFromServer, setTodosFromServer] = useState<Todo[]>([]);
+  const [hasErrors, setHasErrors] = useState(false);
+  const [error, setError] = useState<ErrorSpec | null>(null);
+  const [status, setStatus] = useState<Status>(Status.ALL);
+  const [uncompletedTodosCount, setUncompletedTodosCount] = useState<number>(0);
+  const [isInputDisabled, setIsInputDisabled] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodo, setLoadingTodo] = useState<Todo | null>(null);
+  const [editingTodo, setEditingTodo] = useState<Todo | null>(null);
+
+  const areAllCompleted = uncompletedTodosCount === 0;
+
+  const filteredTodos = useMemo(
+    () => filterTodos(todosFromServer, status),
+    [status, todosFromServer],
+  );
+
+  useEffect(() => {
+    client
+      .get<Todo[]>(ADDED_URL)
+      .then((todos) => {
+        setTodosFromServer(todos);
+
+        if (todos.length === 0) {
+          setError(ErrorSpec.EMPTY_TITLE);
+        }
+      })
+      .catch(() => setError(ErrorSpec.NOT_LOADED));
+  }, []);
+
+  useEffect(() => {
+    const completedCount = todosFromServer.filter(
+      (todo) => !todo.completed,
+    ).length;
+
+    setUncompletedTodosCount(completedCount);
+  }, [todosFromServer, status, filteredTodos]);
+
+  const handleStatus = (newStatus: Status) => {
+    setStatus(newStatus);
+  };
+
+  const cleanErrors = () => {
+    setHasErrors(false);
+  };
+
+  const addTodo = (title: string) => {
+    setIsInputDisabled(true);
+    const preparedTitle = title.trim();
+
+    if (!preparedTitle) {
+      setError(ErrorSpec.EMPTY_TITLE);
+      setHasErrors(true);
+      setIsInputDisabled(false);
+
+      return;
+    }
+
+    const newTodo: Todo = {
+      id: 0,
+      userId: USER_ID,
+      title,
+      completed: false,
+    };
+
+    TodoService.createTodo(newTodo)
+      .then((responce) => {
+        setTempTodo(responce);
+        setTodosFromServer([...todosFromServer, responce]);
+      })
+      .catch(() => {
+        setError(ErrorSpec.NOT_ADDED);
+        setIsInputDisabled(false);
+      })
+      .finally(() => {
+        setIsInputDisabled(false);
+        setTempTodo(null);
+      });
+  };
+
+  const removeTodo = (id: number) => {
+    setIsInputDisabled(true);
+    setLoadingTodo(filteredTodos.find((todo) => todo.id === id) || null);
+
+    TodoService.deleteTodo(id)
+      .then(() => setLoadingTodo(null))
+      .catch(() => {
+        setError(ErrorSpec.NOT_DELETED);
+        setIsInputDisabled(false);
+      })
+      .finally(() => {
+        setTodosFromServer((prev) => prev.filter((todo) => todo.id !== id));
+        setIsInputDisabled(false);
+      });
+  };
+
+  const clearCompleted = () => {
+    filteredTodos.forEach((element) => {
+      if (element.completed) {
+        removeTodo(element.id);
+      }
+    });
+    setUncompletedTodosCount(0);
```

Comment:
requests for deletion may fall and therefore you cannot know for sure that this value will be `0`